File: packages/sdk-ts/src/endpoint-verify.ts
// Swarmwage Agent SDK — endpoint ownership proof
// License: MIT
//
// Wave-2a fix for the squat attack: a malicious seller can publish a
// listing whose `endpoint` URL points at a third party's server (a
// competitor's hire endpoint, an internal admin panel, an SSRF target).
// Wave 1.5 already blocks the obvious SSRF targets at validation time;
// this module closes the structural variant: the registry now demands a
// nonce-signed proof from the endpoint owner that they control the same
// `agent_id` the listing claims. A squatter cannot satisfy the challenge
// because they don't hold the third party's private key.
//
// Wire format (request/response of `GET <endpoint>/.well-known/swarmwage-verify`):
//
//   GET /.well-known/swarmwage-verify?nonce=<opaque>
//   →  200 application/json
//      { "agent_id": "0x…", "nonce": "<echoed>", "signature": "0x…" }
//
// The signature is produced over the canonical payload `{ agent_id, nonce }`
// using the same `signTypedPayload` flow used for listings + receipts
// (stable-key JSON → keccak256 → eth_personalSign). The registry verifies
// it against the listing's `agent_id` via `verifyTypedPayload`.

import type { AgentId, Hex } from "./types.js";

/** Path the registry challenges; sellers serve their proof at this URL. */
export const ENDPOINT_VERIFY_PATH = "/.well-known/swarmwage-verify";

/** Response body schema for the verify endpoint. */
export interface EndpointVerifyResponse {
  agent_id: AgentId;
  nonce: string;
  signature: Hex;
}

/**
 * Build a signed verify response for the well-known endpoint. Callers
 * supply their own `signTypedPayload` (most sellers already have one
 * locally; the SDK's `createWallet` also provides it). The function is
 * framework-agnostic — wrap it in your Hono / Express / Fastify handler.
 *
 * @example
 *   app.get(ENDPOINT_VERIFY_PATH, async (c) => {
 *     const nonce = c.req.query("nonce");
 *     if (!nonce || nonce.length < 8 || nonce.length > 128) {
 *       return c.json({ error: "Invalid or missing nonce" }, 400);
 *     }
 *     return c.json(await signEndpointVerify(agentId, nonce, signTypedPayload));
 *   });
 */
export async function signEndpointVerify(
  agentId: AgentId,
  nonce: string,
  signTypedPayload: (payload: object) => Promise<Hex>,
): Promise<EndpointVerifyResponse> {
  const signature = await signTypedPayload({
    agent_id: agentId,
    nonce,
  });
  return { agent_id: agentId, nonce, signature };
}

File: packages/sdk-py/src/swarmwage/endpoint_verify.py
# Swarmwage Agent SDK (Python) — endpoint ownership proof
# Mirrors packages/sdk-ts/src/endpoint-verify.ts
# License: MIT
#
# Wave-2a registry hardening: a malicious seller can publish a listing
# whose `endpoint` URL points at a third party's server (a competitor,
# an internal admin panel, an SSRF target). The registry challenges the
# endpoint over the well-known path below; the seller must answer with
# a signature over the same `agent_id` it claims in the listing. A
# squatter cannot satisfy the challenge — they don't hold the third
# party's private key.

from __future__ import annotations

from typing import Any

from ._signing import sign_typed_payload

# Path the registry challenges; sellers serve their proof here.
ENDPOINT_VERIFY_PATH = "/.well-known/swarmwage-verify"


def sign_endpoint_verify(
    *,
    agent_id: str,
    nonce: str,
    seller_private_key: str,
) -> dict[str, Any]:
    """Build a signed verify response for the well-known endpoint.

    Wrap the return value in your Flask / FastAPI / Starlette / aiohttp
    handler::

        from fastapi import FastAPI, Query
        from swarmwage.endpoint_verify import ENDPOINT_VERIFY_PATH, sign_endpoint_verify

        app = FastAPI()

        @app.get(ENDPOINT_VERIFY_PATH)
        def verify(nonce: str = Query(..., min_length=8, max_length=128)):
            return sign_endpoint_verify(
                agent_id=AGENT_ID,
                nonce=nonce,
                seller_private_key=SELLER_KEY,
            )
    """
    payload = {"agent_id": agent_id, "nonce": nonce}
    signature = sign_typed_payload(seller_private_key, payload)
    return {"agent_id": agent_id, "nonce": nonce, "signature": signature}


__all__ = ["ENDPOINT_VERIFY_PATH", "sign_endpoint_verify"]
